NetKet/Machine/activations.hpp

```diff
@@ -0,0 +1,86 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_ACTIVATIONS_HPP
+#define NETKET_ACTIVATIONS_HPP
+
+#include <Eigen/Dense>
+#include <complex>
+#include <iostream>
+#include <random>
+#include <vector>
+
+namespace netket {
+
+/**
+  Abstract class for Activations.
+*/
+class AbstractActivation {
+ public:
+  using VectorType = Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>;
+
+  virtual inline void Activate(const VectorType &Z, VectorType &A) = 0;
```

Comment:
Another minor point: Naming this function `Activate` seems a bit unusual to me, it's like saying "we activate the activation" instead of "we apply the activation to this vector". I'd therefore suggest calling it `Apply` (which is also consistent with `ApplyJacobian`) (or, alternatively, `operator()` so that the activation can be called like a function).

Reply:
I prefer the `operator()` option, maybe returning `A` instead of passing it as a reference (hoping that move constructor is working properly). Otherwise just operator with two arguments is also good.